### PR TITLE
Reorganize check_energy

### PR DESCRIPTION
**Pull request by oksanaguba**
Removing allocates for cpair and rair variables, since they are always set to dry constants, 
in parts that are related to energy calculations.
Removing repeated code in check_energy wrt TE and total water.
Removing calls to strings comparisons in physpkg and in check_energy. 
Removing some if-statements related to waccm (mostly regarding energy routines).
There are such calls (waccm and string comparison) in other places, esp. in parametrizations, leaving them for now.

[bfb]

### CODE REVIEW

**Comment by oksanaguba** [comment]
atm_developer passed cs.status.20220124_215259_z64ors on chrysalis

File: components/eam/src/physics/cam/check_energy.F90
Comment (whannah1) [inline components/eam/src/physics/cam/check_energy.F90:1140]: This comment is a little confusing.

**Review by whannah1 — APPROVED** [review]
I know you didn't ask for my review, but it's so nice to see this cleaned up!

File: components/eam/src/physics/cam/check_energy.F90
Comment (oksanaguba) [reply components/eam/src/physics/cam/check_energy.F90:1140]: This is for future functionality that is not in this PR.

File: components/eam/src/physics/cam/check_energy.F90
Comment (oksanaguba) [reply components/eam/src/physics/cam/check_energy.F90:1227]: Thanks, will reformat now.

**Comment by oksanaguba** [comment]
MMF tests are broken
```
[ac.onguba@chrlogin1 ess]$ ./cs.status.20220218_175256_sjjx23 | grep -i fail
    FAIL ERP_Ld3.ne30pg2_EC30to60E2r2.WCYCL1850.chrysalis_intel.allactive-pioroot1 TPUTCOMP Error: TPUTCOMP: Computation time increase > 5% from baseline
  ERP_Ln9.ne4pg2_ne4pg2.F-MMFXX.chrysalis_intel.eam-mmf_fixed_subcycle (Overall: FAIL) details:
    FAIL ERP_Ln9.ne4pg2_ne4pg2.F-MMFXX.chrysalis_intel.eam-mmf_fixed_subcycle RUN time=8
    FAIL ERS.f19_g16.I1850ELM.chrysalis_intel.elm-betr TPUTCOMP Error: TPUTCOMP: Computation time increase > 5% from baseline
    FAIL ERS.f19_g16.I20TRGSWCNPECACNTBC.chrysalis_intel.elm-eca_f19_g16_I20TRGSWCNPECACNTBC TPUTCOMP Error: TPUTCOMP: Computation time increase > 5% from baseline
    FAIL ERS.f19_g16.I20TRGSWCNPRDCTCBC.chrysalis_intel.elm-ctc_f19_g16_I20TRGSWCNPRDCTCBC TPUTCOMP Error: TPUTCOMP: Computation time increase > 5% from baseline
  ERS_Ln9.ne4pg2_ne4pg2.F-MMF1.chrysalis_intel.eam-mmf_crmout (Overall: FAIL) details:
    FAIL ERS_Ln9.ne4pg2_ne4pg2.F-MMF1.chrysalis_intel.eam-mmf_crmout RUN time=8
  ERS_Ln9.ne4pg2_ne4pg2.F-MMF1-RCEMIP.chrysalis_intel (Overall: FAIL) details:
    FAIL ERS_Ln9.ne4pg2_ne4pg2.F-MMF1-RCEMIP.chrysalis_intel RUN time=7
  ERS_Ln9.ne4pg2_ne4pg2.F-MMFOMP.chrysalis_intel.eam-single_thread (Overall: FAIL) details:
    FAIL ERS_Ln9.ne4pg2_ne4pg2.F-MMFOMP.chrysalis_intel.eam-single_thread RUN time=7
  ERS_Ln9.ne4pg2_ne4pg2.F-MMFXX.chrysalis_intel.eam-mmf_use_VT (Overall: FAIL) details:
    FAIL ERS_Ln9.ne4pg2_ne4pg2.F-MMFXX.chrysalis_intel.eam-mmf_use_VT RUN time=7
  ERS_Ln9_P96x1.ne4pg2_ne4pg2.F-MMFXX.chrysalis_intel.eam-mmf_use_ESMT (Overall: FAIL) details:
    FAIL ERS_Ln9_P96x1.ne4pg2_ne4pg2.F-MMFXX.chrysalis_intel.eam-mmf_use_ESMT RUN time=10
    FAIL PEM_Ln9.ne30pg2_EC30to60E2r2.WCYCL1850.chrysalis_intel TPUTCOMP Error: TPUTCOMP: Computation time increase > 5% from baseline
    FAIL PET_Ln9_PS.ne30pg2_EC30to60E2r2.WCYCL1850.chrysalis_intel.allactive-mach-pet BASELINE master: ERROR BFAIL some baseline files were missing
    FAIL SMS_Ld20.f45_f45.IELMFATES.chrysalis_intel.elm-fates_eca TPUTCOMP Error: TPUTCOMP: Computation time increase > 5% from baseline
    FAIL SMS_Ld2.ne30pg2_r05_EC30to60E2r2.BGCEXP_CNTL_CNPECACNT_1850.chrysalis_intel.elm-bgcexp TPUTCOMP Error: TPUTCOMP: Computation time increase > 5% from baseline
    FAIL SMS_Ld2.ne30pg2_r05_EC30to60E2r2.BGCEXP_CNTL_CNPRDCTC_1850.chrysalis_intel.elm-bgcexp TPUTCOMP Error: TPUTCOMP: Computation time increase > 5% from baseline
    FAIL SMS_Ln5.ne30pg2_EC30to60E2r2.BGCEXP_LNDATM_CNORDCTC_20TR.chrysalis_intel TPUTCOMP Error: TPUTCOMP: Computation time increase > 5% from baseline
  SMS_Ln5.ne4_ne4.F-MMFXX-SCM-ARM97.chrysalis_intel (Overall: FAIL) details:
    FAIL SMS_Ln5.ne4_ne4.F-MMFXX-SCM-ARM97.chrysalis_intel RUN time=3
```

**Comment by oksanaguba** [comment]
Baseline build has a couple of fails (haven't looked)
```
[ac.onguba@chrlogin1 ess]$ ./cs.status.20220217_082031_byh93p | grep -i fail
  PET_Ln9_PS.ne30pg2_EC30to60E2r2.WCYCL1850.chrysalis_intel.allactive-mach-pet (Overall: FAIL) details:
    FAIL PET_Ln9_PS.ne30pg2_EC30to60E2r2.WCYCL1850.chrysalis_intel.allactive-mach-pet RUN time=72
  SMS_Ln5.ne30pg2_EC30to60E2r2.BGCEXP_LNDATM_CNORDCTC_1850.chrysalis_intel (Overall: FAIL) details:
    FAIL SMS_Ln5.ne30pg2_EC30to60E2r2.BGCEXP_LNDATM_CNORDCTC_1850.chrysalis_intel RUN time=1938
```

File: components/eam/src/physics/cam/constituents.F90
Comment (whannah1) [inline components/eam/src/physics/cam/constituents.F90:452]: I don't think it's a good idea to do this. The constituent interface is meant to be generic so that the physics package can decide what constituents need to be added. It's pretty clear that the error you noticed would be from not calling setup_moist_indices() in the MMF physpkg.

The constituent names for the MMF are set in crm_physics.F90. We could move this to the MMF specific physpkg, but I thought it made since to have in crm_physics so it could be modified according to the CRM options (i.e. microphysics and variance tracers). 

Currently, check_energy_register() is called before crm_physics_register() in the MMF physpkg, but there's no reason we couldn't change the order of this so that check_energy_register() can set the indices correctly.

So I think a better strategy would be to have local indices in check_energy.F90 (and wherever else they are needed) and have the values set in check_energy_register() after the call to crm_physics_register().

Comment (singhbalwinder) [reply components/eam/src/physics/cam/constituents.F90:452]: I agree with Walter. I think constituents.F90 provides all the get/set functions for individual species, which keeps it generic.

**Comment by oksanaguba** [comment]
@whannah1 as you said, the indices were not set for MMF. i tried this fix
```
--- a/components/eam/src/physics/crm/physpkg.F90
+++ b/components/eam/src/physics/crm/physpkg.F90
@@ -24,7 +24,7 @@ module physpkg
   use phys_grid,               only: get_ncols_p, print_cost_p, update_cost_p
   use phys_gmean,              only: gmean_mass
   use ppgrid,                  only: begchunk, endchunk, pcols, pver, pverp
-  use constituents,            only: pcnst, cnst_name, cnst_get_ind
+  use constituents,            only: pcnst, cnst_name, cnst_get_ind, setup_moist_indices
   use camsrfexch,              only: cam_out_t, cam_in_t
   use phys_control,            only: phys_do_flux_avg, phys_getopts
   use scamMod,                 only: single_column, scm_crm_mode
@@ -584,6 +584,8 @@ subroutine phys_init( phys_state, phys_tend, pbuf2d, cam_out )
   ! diag_init makes addfld calls for dyn fields that are output from the physics decomp
   call diag_init()
 
+  call setup_moist_indices()
+
   call check_energy_init()
```
in 1 MMF test and it worked.

I am not exactly following why you suggest to have indices set in check_energy. indices, as tracer names, remain the same during a simulation. the change i proposed just removes string comparisons (which are slow), indices then should belong to constituents mod? the indices themselves are protected, so other code cannot change them arbitrarily.

thanks!

**Comment by whannah1** [comment]
> I am not exactly following why you suggest to have indices set in check_energy. indices, as tracer names, remain the same during a simulation. the change i proposed just removes string comparisons (which are slow), indices then should belong to constituents mod? the indices themselves are protected, so other code cannot change them arbitrarily.

The tracer names/indices can change depending on the configuration, and since we want the constituents interface to remain generic for any list of water tracers. Having setup_moist_indices() defined the way it is in constituents.F90 goes against this, so it is not how we should address this. 

Having the moist indices be defined and set locally in a module that requires them (check_energy.F90 in this case) seems to be a much better way of handling this issue. I would rather take this approach than use the method that is currently suggested in this PR.

**Comment by oksanaguba** [comment]
The current design assumes fixed names for water forms, the names are hardcoded as strings all over the code. So, what is done in the PR follows the same design. 

The only thing that this PR does is that it removes 100x string comparisons per time step. In future, we could think about a case where water forms change names, but even then, indices (which are set in the current design based on water forms' names) will remain constant during simulation. I do not see why making each module set their own (same) water form indices is a better design than having them set in constituents module.

Maybe there is a confusion about how indices are set in this PR:
```
subroutine setup_moist_indices()

   implicit none

   call cnst_get_ind('CLDICE', icldice, abrtf=.false.)
   call cnst_get_ind('CLDLIQ', icldliq, abrtf=.false.)
   call cnst_get_ind('RAINQM', irain, abrtf=.false.)
   call cnst_get_ind('SNOWQM', isnow, abrtf=.false.)

 end subroutine setup_moist_indices
```

The indices are set based on water form names (just like in master), they are not hardcoded.

**Comment by singhbalwinder** [comment]
> The only thing that this PR does is that it removes 100x string comparisons per time step

I thought these string comparisons should stay in the model initialization step (`init` routine) as these indices should not change during the time stepping. Isn't that the case? There may be places in the code where these are done during the time stepping but we should clean those up  and move them to the init routine as we encounter them. This will avoid these string comparisons being done during the time stepping.

**Comment by oksanaguba** [comment]
@singhbalwinder "I thought these string comparisons should stay in the model initialization step (init routine) as these indices should not change during the time stepping. Isn't that the case?" -- the PR does this for check_energy and for physpkg . Note that string comparisons remain in many files of parametrizations (i consider that out of scope of this PR).

**Comment by whannah1** [comment]
@oksanaguba I still think the approach here is bad practice. Or perhaps I should say that it propagates a bad practice from the older code. Either way, I don't want to hold things up. If my concerns come to fruition about needing a more flexible way to refer to constituents across different physics options, then we can deal with it later.

**Comment by oksanaguba** [comment]
Thanks @whannah1 and @singhbalwinder . Yes, we should reconsider this is there is a need.

Regarding @singhbalwinder 's question above about indices set in init -- in the original code, looking, for example, at check_energy change routine, I see that indices are searched each call of the routine: 
```
  subroutine check_energy_chng(state, tend, name, nstep, ztodt,        &
       flx_vap, flx_cnd, flx_ice, flx_sen)
...
    call cnst_get_ind('CLDICE', ixcldice, abrtf=.false.)
    call cnst_get_ind('CLDLIQ', ixcldliq, abrtf=.false.)
    call cnst_get_ind('RAINQM', ixrain, abrtf=.false.)
    call cnst_get_ind('SNOWQM', ixsnow, abrtf=.false.)
```

**Comment by oksanaguba** [comment]
These tests now pass
```
test=ERP_Ln9.ne4pg2_ne4pg2.F-MMFXX.chrysalis_intel.eam-mmf_fixed_subcycle
${brr}/cime/scripts/create_test    $suff -c  -b master ${test}
test=ERS_Ln9.ne4pg2_ne4pg2.F-MMF1.chrysalis_intel.eam-mmf_crmout
${brr}/cime/scripts/create_test    $suff -c  -b master ${test}
test=ERS_Ln9.ne4pg2_ne4pg2.F-MMF1-RCEMIP.chrysalis_intel
${brr}/cime/scripts/create_test    $suff -c  -b master ${test}
test=ERS_Ln9.ne4pg2_ne4pg2.F-MMFOMP.chrysalis_intel.eam-single_thread
${brr}/cime/scripts/create_test    $suff -c  -b master ${test}
test=ERS_Ln9.ne4pg2_ne4pg2.F-MMFXX.chrysalis_intel.eam-mmf_use_VT
${brr}/cime/scripts/create_test    $suff -c  -b master ${test}
test=ERS_Ln9_P96x1.ne4pg2_ne4pg2.F-MMFXX.chrysalis_intel.eam-mmf_use_ESMT
${brr}/cime/scripts/create_test    $suff -c  -b master ${test}
test=SMS_Ln5.ne4_ne4.F-MMFXX-SCM-ARM97.chrysalis_intel
${brr}/cime/scripts/create_test    $suff -c  -b master ${test}
```

**Comment by singhbalwinder** [comment]
Thanks @oksanaguba. You are right, currently we are getting the indices every time step. We can move this code to the init routine and declare the indices as global vars for this module. I think this is a better approach than creating a new subroutine (`setup_mosit_indices`) to avoid these string comparisons at each time step.

**Comment by oksanaguba** [comment]
@singhbalwinder are you suggesting moving the code that inits indices into each module file into each (if available) init routine? if so, I am not sure how that is better than init-ing indices once in a more general init (this is what's done right now). thanks.

**Comment by singhbalwinder** [comment]
@oksanaguba : I am suggesting to do this only for check_energy.F90 file (not all the module files in EAM). This will avoid the string comparisons (which I agree are slow and probably error prone as well) during the time stepping of check_energy.F90. This way, we don't need to add and call `setup_moist_indices` routine, which I think may led to addition of other such routines in the future. 

This is my personal opinion and just a suggestion. I do not want to hold up this PR as it includes some other nice code cleanup changes.

**Comment by rljacob** [comment]
@wlin7 please review.

**Comment by oksanaguba** [comment]
@rljacob i'd like to merge it today after re-running tests again. @whannah1 and @singhbalwinder had some useful comments about this PR, i believe they are ok with it as is for now.

**Comment by rljacob** [comment]
sure go ahead.

**Comment by oksanaguba** [comment]
```
[ac.onguba@chrlogin1 acme-integrator]$ ~/ess/cs.status.20220513_160549_biyswm | grep -i fail
    FAIL ERP_Ln9.ne4_ne4.FAQP.chrysalis_intel BASELINE master: ERROR BFAIL some baseline files were missing
    FAIL ERP_Ln9.ne4_ne4.FRCE.chrysalis_intel TPUTCOMP Error: TPUTCOMP: Computation time increase > 5% from baseline
    FAIL ERS_Ld5.ne4_oQU240.F2010.chrysalis_intel.eam-rrtmgpxx TPUTCOMP Error: TPUTCOMP: Computation time increase > 5% from baseline
    FAIL SMS_Ln5.ne30pg2_r05_EC30to60E2r2.BGCEXP_LNDATM_CNPRDCTC_20TR.chrysalis_intel TPUTCOMP Error: TPUTCOMP: Computation time increase > 5% from baseline
    FAIL SMS.ne4_oQU240.F2010.chrysalis_intel.eam-preqx_ftype1 MEMCOMP Error: Memory usage increase >5% from baseline's 2015.270000 to 2116.370000
    FAIL SMS.ne4_oQU240.F2010.chrysalis_intel.eam-preqx_ftype4 MEMCOMP Error: Memory usage increase >5% from baseline's 2015.280000 to 2118.380000
    FAIL SMS.ne4_oQU240.F2010.chrysalis_intel.eam-thetahy_ftype0 MEMCOMP Error: Memory usage increase >5% from baseline's 2032.670000 to 2134.860000
    FAIL SMS.ne4_oQU240.F2010.chrysalis_intel.eam-thetanh_ftype0 MEMCOMP Error: Memory usage increase >5% from baseline's 2019.720000 to 2120.820000
```

**Comment by oksanaguba** [comment]
in next

**Comment by oksanaguba** [comment]
There is a fail on cdash, related to this merge:
https://my.cdash.org/test/55655779
```
Test: SMS_D_Ld1.ne30pg2_r05_EC30to60E2r2.WCYCL1850.anvil_gnu (Failed)
Build: e3sm_integration_next_gnu[ (anvil) ](https://my.cdash.org/viewSite.php?siteid=18594)on 2022-05-15 11:41:09
Test Details: Completed (FAIL)
```

I ran manually from some clone that has this PR:
```
[ac.onguba@chrlogin2 run]$ pwd
/home/ac.onguba/es/SMS_D_Ld1.ne30pg2_r05_EC30to60E2r2.WCYCL1850.anvil_gnu.C.20220516_163005_twgw30/run
[ac.onguba@chrlogin2 run]$ head -n 5 ../bld/GIT_LOG
c14fb54c1cac63973b9bf50bbdb329f8378c02d4 Merge remote-tracking branch 'OG/og/cleanup' into HEAD
8b12fc569cce6269e586cd345ab084ba5399fee3 Merge branch 'whannah/mmf/rename-compsets' (#4837)
ed696b52436de6eae0bbcad821e3170ebce08a15 Merge branch 'maltrud/ocean/fix-BGC-diags' (PR #4944)
013fe19a48fcd5221038a2a022d5298b072fccb7 Merge branch 'jonbob/coupler/fix-bgcbudget-restart' (PR #4934)
872c03755298b08151839b1d4757e200faddf8f7 Merge branch 'matthewhoffman/mpas/stream_read_input_only' (PR #4929)
```
and from next as of Friday, right after the merge:
```
[ac.onguba@chrlogin2 run]$ pwd
/home/ac.onguba/es/SMS_D_Ld1.ne30pg2_r05_EC30to60E2r2.WCYCL1850.anvil_gnu.C.20220516_183214_kqhd2m/run
[ac.onguba@chrlogin2 run]$ head -n 5 ../bld/GIT_LOG
caf9ddbee9e0a2485cc9012f82624b082818c382 Merge remote-tracking branch 'OG/og/cleanup' (PR #4753)
012fcdd659f646deb0a1980fbca468c6171cc32b Merge branch 'azamat/anvil/trigrid-bgc-pes' into next (PR #4950)
f4fd314ce9979ae5867126f2a15243a84a55bbc7 Merge branch 'azamat/anvil/trigrid-bgc-pes' into next (PR #4950)
3a3c523607441f830bf00bf0f3c3fa466d47ba26 Merge branch 'mt5555-dtime-updates' into next (PR #4947)
ac9b3cc03336e41172babeaf12fbaa7c846d583a Merge branch 'maltrud/ocean/fix-BGC-diags' into next (PR #4944)
```
Both runs finished without the crash from nightlies.
```

**Pull request merged**
